# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/builtin/TestingFunctions.cpp
# Commit: 2da058734d5c
# Full Hash: 2da058734d5c3f3230390c75eee4032d9f1de4f8
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2021-09-09 15:22:41
# Regressor Bug: 1720422
# File Overlap Count: 2
# Description:
#   Bug 1720422 - Move typed array unwrap API to templatized version r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D119856
# ==============================================================================

diff -r 4791d6790af7 -r 2da058734d5c js/src/builtin/TestingFunctions.cpp
--- a/js/src/builtin/TestingFunctions.cpp	Wed Sep 08 23:46:57 2021 +0000
+++ b/js/src/builtin/TestingFunctions.cpp	Wed Sep 08 23:46:58 2021 +0000
@@ -7005,15 +7005,22 @@
   }
   array->ensureDenseInitializedLength(0, 2);
 
+  JSObject* obj = args[1].isObject() ? &args[1].toObject() : nullptr;
+  auto view = JS::TypedArray<Scalar::Uint8>::unwrap(obj);
+  if (!view) {
+    ReportUsageErrorASCII(cx, callee, "Second argument must be a Uint8Array");
+    return false;
+  }
   size_t length;
   bool isSharedMemory;
-  uint8_t* data;
-  if (!args[1].isObject() ||
-      !JS_GetObjectAsUint8Array(&args[1].toObject(), &length, &isSharedMemory,
-                                &data) ||
-      isSharedMemory ||  // excluded views of SharedArrayBuffers
+  JS::AutoCheckCannotGC nogc(cx);
+  uint8_t* data = view.getLengthAndData(&length, &isSharedMemory, nogc);
+
+  if (isSharedMemory ||  // excluded views of SharedArrayBuffers
       !data) {           // exclude views of detached ArrayBuffers
-    ReportUsageErrorASCII(cx, callee, "Second argument must be a Uint8Array");
+    ReportUsageErrorASCII(
+        cx, callee,
+        "Second argument must be an unshared, non-detached Uint8Array");
     return false;
   }
 